apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -140,66 +176,90 @@ const DanceAiModal: React.FunctionComponent<DanceAiProps> = ({onClose}) => {
     }
   };
 
-  const handleResultsClick = () => {
-    setMode(Mode.RESULTS);
-  };
-
-  const handleResultsFinalClick = () => {
-    setMode(Mode.RESULTS_FINAL);
-  };
-
   const handleGenerateClick = () => {
     startAi();
     setMode(Mode.GENERATING);
   };
 
-  const handleRegenerateClick = () => {
-    // Reset state
-    setTypingDone(false);
-    setResultJson('');
-    setShowPreview(false);
+  const handleStartOverClick = () => {
+    setInputs([]);
+    setCurrentInputSlot(0);
     setProcessingDone(false);
-    setGeneratingNodesDone(false);
-    setGeneratingDone(false);
+    setGeneratingStep([0, 0]);
+    setMode(Mode.SELECT_INPUTS);
+  };
 
+  const handleRegenerateClick = () => {
+    setProcessingDone(false);
+    setGeneratingStep([0, 0]);
     handleGenerateClick();
   };
 
   const handleProcessClick = () => {
     setMode(Mode.PROCESSING);
     setTimeout(() => {
-      setProcessingDone(true);
+      handleGenerateClick();
     }, 4000);
   };
 
+  const handleExplanationClick = () => {
+    setMode(Mode.EXPLANATION);
+  };
+
+  const handleUseClick = () => {
+    currentAiModalField?.setValue(
+      JSON.stringify({
+        inputs,
+        ...generatedEffects.current[BAD_GENERATED_RESULTS_COUNT].results,
+      })
+    );
+    dispatch(setCurrentAiModalField(undefined));
+  };
+
+  const handleLeaveExplanation = () => {
+    setMode(Mode.RESULTS);
+  };
+
+  // Animate through the generating process.
+  useInterval(
+    () => {
+      if (mode === Mode.GENERATING) {
+        // We do a deep copy into a new array to ensure that
+        // a re-render is triggered at the end of this work.
+        const currentGeneratingStep: GeneratingStep = [...generatingStep];
+
+        if (currentGeneratingStep[1] < GENERATING_SUBSTEP_COUNT - 1) {
```

Comment:
Similarly (for readability), could this be something like
```
type GeneratingStep = {
  currentStep: number;
  currentSubstep: number;
}
```
so it's more obvious what each of the fields refers to?

Reply:
tbh I'm struggling to understand what the substep represents...I see the animations sort of "flicker" part way though each step in your demo video? Mind explaining a bit what the substep is for?

Reply:
This was designed to let us animate through the generating process.  The step represents a major change, in this case showing a new preview, and substep is for whatever happens sequentially inside of each step.  Right now, it's indeed just two steps: in the first step, the bot head stays normal, and in the second step, the bot head shows its reaction.  But during development, and indeed soon, we might have additional steps, so that we can show a score in a separate substep to the bot head reacting, etc.

Reply:
got it, thanks!

Reply:
Good call on the type; now using one for generating progress.